src/kmeans.jl

```diff
@@ -292,37 +285,33 @@ end
 #  (specific to the case where samples are weighted)
 #
 function update_centers!(
-    x::Matrix{T},                   # in: sample matrix (d x n)
-    weights::Vector{T},             # in: sample weights (n)
-    assignments::Vector{Int},       # in: assignments (n)
-    to_update::Vector{Bool},        # in: whether a center needs update (k)
-    centers::Matrix{T},             # out: updated centers (d x k)
-    cweights::Vector                # out: updated cluster weights (k)
-) where T<:AbstractFloat
-
-    d::Int = size(x, 1)
-    n::Int = size(x, 2)
-    k::Int = size(centers, 2)
+    X::AbstractMatrix{<:Real},       # in: sample matrix (d x n)
+    weights::AbstractVector{<:Real}, # in: sample weights (n)
+    assignments::Vector{Int},        # in: assignments (n)
+    to_update::Vector{Bool},         # in: whether a center needs update (k)
+    centers::AbstractMatrix{<:Real}, # out: updated centers (d x k)
+    cweights::Vector{Float64})       # out: updated cluster weights (k)
+
+    (d, n), k = size(X), size(centers, 2)
 
     # initialize center weights
     cweights[to_update] .= 0.0
 
     # accumulate columns
     # accumulate_cols_u!(centers, cweights, x, assignments, weights, to_update)
-    for j = 1 : n
+    for j = 1:n
         @inbounds wj = weights[j]
-
         if wj > 0
             @inbounds cj = assignments[j]
             1 <= cj <= k || error("assignment out of boundary.")
 
             if to_update[cj]
                 rj = view(centers, :, cj)
-                xj = view(x, :, j)
+                xj = view(X, :, j)
                 if cweights[cj] > 0
                     @inbounds rj .+= xj * wj
                 else
-                    @inbounds rj .= xj * wj
+                    rj .= xj * wj
```

Comment:
Are you sure `@inbounds` is not required here?

Reply:
I put it back